nsqd/gossip_test.go

```diff
@@ -0,0 +1,141 @@
+package nsqd
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+)
+
+var gossipNotifyCh = make(chan struct{}, 20)
+var (
+	converging     = false
+	convergingLock = &sync.Mutex{}
+)
+
+func TestGossip(t *testing.T) {
+	var nsqds []*NSQD
+	var seedNodeAddr string
+	var seedNode *NSQD
+	var tcpPorts []int
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	gossipNotify = func() {
+		convergingLock.Lock()
+		defer convergingLock.Unlock()
+		if converging {
+			gossipNotifyCh <- struct{}{}
+		}
+	}
+
+	num := 3
+	for i := 0; i < num; i++ {
+		var nsqd *NSQD
+
+		// find an open port
+		tmpl, err := net.Listen("tcp", hostname+":0")
+		equal(t, err, nil)
+		addr := tmpl.Addr().(*net.TCPAddr)
+		tmpl.Close()
+
+		opts := NewNSQDOptions()
+		opts.ID = int64(i)
+		opts.Logger = newTestLogger(t)
+		opts.GossipAddress = addr.String()
+		if seedNode != nil {
+			sn := net.JoinHostPort(seedNodeAddr, fmt.Sprint(seedNode.serf.Memberlist().LocalNode().Port))
+			opts.SeedNodeAddresses = []string{sn}
+		}
+		tcpAddr, _, nsqd := mustStartNSQD(opts)
+		defer func(n *NSQD) {
```

Comment:
this new defer shouldn't be necessary, I don't think it was the cause of your "test never ending" issue